patterns/interpreter.md

```diff
@@ -0,0 +1,317 @@
+language, then define a grammar for this language and
+design interpreter solving problem instances.
+
+## Motivation
+
+Imagine that our work is translating simple mathematical expressions into
+[assembly language](https://en.wikipedia.org/wiki/Assembly_language)
+(more simple and low level programming language).
+For simplicity, our expressions consists of ten digits `0`,...,`9`,
+four operations `+, -, /, *` and a pair of parenthesis `(, )`.
+For example, expression `2 + 4` could be translated into
+
+```ignore
+mov eax, 2
+mov ebx, 4
+add eax, ebx
+```
+
+Our goal is to automate translating into assembly instructions
+using the Interpreter design pattern. In other words, we want simply
+provide the Interpreter with an expression and get Assembly
+language output. For example
+
+```rust, ignore
+x.interpret("7+3*(2-1)", &output);
+```
+
+## Context Free Grammars
+
+Basicaly, a
+[Context Free Grammar](https://en.wikipedia.org/wiki/Context-free_grammar)
+(CFG) describes (or defines) all possible strings
+over some finite set of symbols, formally called alphabet.
+These strings usually has some well defined structure.
+For example, strings of all balanced parentheses
+
+```ignore
+(), (()), ()(), ()(()),..
+```
+
+Formally a CFG consists of three sets: terminal symbols,
+nonterminal symbols, a start (nonterminal) symbol,
+and production rules.
+
+In the previous example we have two terminal symbols `(` and `)`.
+We define a single nonterminal (at the same time the start)
+symbol `S` and priduction rules as
+
+```ignore
+S -> SS
+S ->(S)
+S ->()
+```
+
+So, in order to generate a string with balanced parentheses we
+repeatedly apply production rules.
+For example, to derive `()()` we apply the following rules
+
+```ignore
+S -> SS -> ()(S) -> ()()
+```
+
+Next, let's define a context free grammar
+for a set of expressions over `0,...,9, +,-,*,/,(,)`, where
+
+- terminal symbols: `0,...,9, +,-,*,/,(,)`
+- nonterminal symbols: `exp, term, factor`
+- start symbol is `exp`
+- and the following are production rules
+
+```ignore
+exp -> exp + term
+exp -> exp - term
+exp -> term
+term -> term * factor
+term -> term / factor
+term -> factor
+factor -> ( exp )
+factor -> 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
+```
+
+In fact this grammar should be further transformed
+depending on what we are going to do with the grammar.
+For example, we might need to remove left recursion.
+For more details please see
+[Compilers:Principles,Techniques, and Tools
+](https://en.wikipedia.org/wiki/Compilers:_Principles,_Techniques,_and_Tools)
+(aka Dragon Book).
+
+## Solution 1
+
+Our first approach is a standard one, simple implementation of
+a recursive descent parser. The following code
+doesn't have `struct` abstraction in order to keep code short.
+The code panics when expression is syntactically wrong
+(unbalanced parentheses or missing digit/operator for example).
+
+```rust
+fn token(input: &[u8], cur: usize) -> char {
```

Comment:
first question that came to my mind: "what's cur?"

Reply:
It stands for current? Maybe we should call this `index` or something like that?
In general I like extended nouns, but this might be only my personal preference.

Reply:
> first question that came to my mind: "what's cur?"
`cur` is the index pointing to the current symbol.
Yeah, we could rename it for readability.